Add test to verify downgrade protection

simo5 opened the pull request:
### Description
RFC8446 mandates that TLS 1.3 Servers set special values in the last
8 bytes of the ServerHello Random buffer when replying to TLS 1.2
and below clients.
This allows client to detect that a downgrade happened.

This test checks that server support the correct downgrade protection
value when the client negotiates TLS versions 1.0, 1.1, 1.2


### Motivation and Context
Fixes #245 

### Checklist
- [x] I have read the [CONTRIBUTING.md](https://github.com/tomato42/tlsfuzzer/blob/master/CONTRIBUTING.md) document and my PR follows [change requirements](https://github.com/tomato42/tlsfuzzer/blob/master/CONTRIBUTING.md#change-requirements) therein
- [x] the changes are also reflected in documentation and code comments
- [x] all new and existing tests pass (see Travis CI results)
- [x] [test script checklist](https://github.com/tomato42/tlsfuzzer/wiki/Test-script-checklist) was followed for new scripts
- [x] new test script added to `tlslite-ng.json` and `tlslite-ng-random-subset.json`
- [x] new and modified scripts were ran against popular TLS implementations:
  - [x] OpenSSL
  - [ ] NSS
  - [x] GnuTLS
- [x] required version of tlslite-ng updated in requirements.txt and README.md

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/tomato42/tlsfuzzer/479)
<!-- Reviewable:end -->

simo5 commented:
Note the only public server I found that seem to support downgrade protection is https://enabled.tls13.com from this list:
https://github.com/tlswg/tls13-spec/wiki/Implementations#test-servers

The other servers either do not allow downgrade at all or return regular random values without the prescribed downgrade protection.

simo5 commented:
Please advise on how to deal with codeclimate issue with 5 arguments to the __init__ function.
Creating a whole new class just to add downgrade protection and fool the __init__ check seems stupid, but that's the only idea I came up with.

tomato42 commented:
https://github.com/tomato42/tlslite-ng/pull/324 should make tlslite-ng follow RFC 8446 on server side

simo5 commented:
Changed code to pass a --server-max-protocol option, which triggers tests based on the actual servere returned version.
Negative tests are also performed to confirm the server do not return the sentinel when it shouldn't.

simo5 commented:
Fixed unused import

tomato42 commented:
> We are testing downraed from TLS1.3 to TLS1.2, it's intentionally a 1.2 handshake.

but it applies only to servers that actually have TLS 1.3 supported and enabled – if the server does not support TLS 1.3 it should not send the sentinel, so a sanity test case isn't really checking sanity of the environment

> Do we care ? If so, what should we do ? Error out ?

yes, we care – typos are easy to make and would remain invisible

raising simple ValueError exception is sufficient

> I do not understand how you can make conflicting tests in the same test runs.

you can specify different configurations for the server:
https://github.com/tomato42/tlsfuzzer/blob/1498cbb33b2400515bfa3cbc6d1c6857ab47ea7f/tests/tlslite-ng.json#L294-L298

https://github.com/tomato42/tlsfuzzer/blob/1498cbb33b2400515bfa3cbc6d1c6857ab47ea7f/tests/tlslite-ng.json#L339-L342

> Not sure what's wrong ?

the test needs to run against a `tls.py server --max-ver tls1.2` too

> It doesn't.

then please add a test case that does set the max version to TLS 1.3 and negotiates TLS 1.3 connection

simo5 commented:
On Tue, 2018-12-11 at 15:33 +0000, Hubert Kario wrote:
> > We are testing downraed from TLS1.3 to TLS1.2, it's intentionally a
> > 1.2 handshake.
> 
> but it applies only to servers that actually have TLS 1.3 supported
> and enabled – if the server does not support TLS 1.3 it should not
> send the sentinel, so a sanity test case isn't really checking sanity
> of the environment

The sanity test does not check the sentinel at all because it doesn't
set server_max_protocol

> > Do we care ? If so, what should we do ? Error out ?
> 
> yes, we care – typos are easy to make and would remain invisible
> 
> raising simple ValueError exception is sufficient

ok

> > I do not understand how you can make conflicting tests in the same
> > test runs.
> 
> you can specify different configurations for the server:
> 
https://github.com/tomato42/tlsfuzzer/blob/1498cbb33b2400515bfa3cbc6d1c6857ab47ea7f/tests/tlslite-ng.json#L294-L298
> 
> 
https://github.com/tomato42/tlsfuzzer/blob/1498cbb33b2400515bfa3cbc6d1c6857ab47ea7f/tests/tlslite-ng.json#L339-L342

ok

> > Not sure what's wrong ?
> 
> the test needs to run against a `tls.py server --max-ver tls1.2` too

ok

> > It doesn't.
> 
> then please add a test case that does set the max version to TLS 1.3
> and negotiates TLS 1.3 connection

ok

tomato42 commented:
>  > but it applies only to servers that actually have TLS 1.3 supported and enabled – if the server does not support TLS 1.3 it should not send the sentinel, so a sanity test case isn't really checking sanity of the environment

> The sanity test does not check the sentinel at all because it doesn't set server_max_protocol

but the only way to verify that the script wasn't provided incorrect information about whether the server does or does not support TLS 1.3 is to actually negotiate it and that's missing now

simo5 commented:
I added unit tests, for various combinations, I also changed the functional test to assume a TLSv1.3 server by default. In that case the "sanity" connection uses TLSv1.3.
Finally I added a tls1.2 forced server configuration and thest in the json files.
This should address all your requests.

tomato42 commented:
This pull request **introduces 2 alerts** when merging e349ec5339119a5bb17f060d822329bc431002f4 into 1498cbb33b2400515bfa3cbc6d1c6857ab47ea7f - [view on LGTM.com](https://lgtm.com/projects/g/tomato42/tlsfuzzer/rev/pr-38c435a92a9eebe9334f0aab238833b1a3051b31)

**new alerts:**

* 2 for Testing equality to None

---

*Comment posted by [LGTM.com](https://lgtm.com)*

tomato42 commented:
looks good, thank you!
